clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/lifecycle/InstanceTerminationLifecycleAgent.java

```diff
@@ -0,0 +1,220 @@
+                                           InstanceTerminationConfigurationProperties properties,
+                                           Provider<AwsEurekaSupport> discoverySupport) {
+    this.objectMapper = objectMapper;
+    this.amazonClientProvider = amazonClientProvider;
+    this.accountCredentialsProvider = accountCredentialsProvider;
+    this.properties = properties;
+    this.discoverySupport = discoverySupport;
+
+    Set<? extends AccountCredentials> accountCredentials = accountCredentialsProvider.getAll();
+    this.queueARN = new ARN(accountCredentials, properties.getQueueARN());
+  }
+
+  @Override
+  public String getAgentType() {
+    return queueARN.account.getName() + "/" + queueARN.region + "/" + InstanceTerminationLifecycleAgent.class.getSimpleName();
+  }
+
+  @Override
+  public String getProviderName() {
+    return AwsProvider.PROVIDER_NAME;
+  }
+
+  @Override
+  public long getPollIntervalMillis() {
+    return TimeUnit.SECONDS.toMillis(30);
+  }
+
+  @Override
+  public long getTimeoutMillis() {
+    return -1;
+  }
+
+  @Override
+  public void run() {
+    AmazonSQS amazonSQS = amazonClientProvider.getAmazonSQS(queueARN.account, queueARN.region);
+
+    Set<String> allAccountIds = accountCredentialsProvider.getAll()
+      .stream()
+      .map(AccountCredentials::getAccountId)
+      .filter(a -> a != null)
+      .collect(Collectors.toSet());
+
+    this.queueId = ensureQueueExists(amazonSQS, queueARN, properties.getSourceARN(), allAccountIds);
+
+    AtomicInteger messagesProcessed = new AtomicInteger(0);
+    while (messagesProcessed.get() < properties.getMaxMessagesPerCycle()) {
+      ReceiveMessageResult receiveMessageResult = amazonSQS.receiveMessage(
+        new ReceiveMessageRequest(queueId)
+          .withMaxNumberOfMessages(AWS_MAX_NUMBER_OF_MESSAGES)
+          .withVisibilityTimeout(properties.getVisibilityTimeout())
+          .withWaitTimeSeconds(properties.getWaitTimeSeconds())
+      );
+
+      receiveMessageResult.getMessages().forEach(message -> {
+        try {
+          LifecycleMessage lifecycleMessage = objectMapper.readValue(message.getBody(), LifecycleMessage.class);
+
+          if (SUPPORTED_LIFECYCLE_TRANSITION.equalsIgnoreCase(lifecycleMessage.lifecycleTransition)) {
+            Task originalTask = TaskRepository.threadLocalTask.get();
+            try {
+              TaskRepository.threadLocalTask.set(
+                Optional.ofNullable(originalTask).orElse(new DefaultTask(InstanceTerminationLifecycleAgent.class.getSimpleName()))
+              );
+              handleMessage(lifecycleMessage, TaskRepository.threadLocalTask.get());
+            } finally {
+              TaskRepository.threadLocalTask.set(originalTask);
+            }
+          }
+        } catch (IOException e) {
+          log.error("Unable to convert NotificationMessage (body: {})", message.getBody(), e);
+        }
+
+        deleteMessage(amazonSQS, queueId, message);
```

Comment:
We don't want to keep processing messages that fail, otherwise we'll likely fall behind during larger / concurrent deployments. Letting the message go unprocessed isn't necessarily bad, since the default behavior will just be that the instance falls out of discovery a little slower (which is what the current behavior is).